File: data/src/main/java/js/task/data/DataProvider.java
package js.task.data;

import java.util.List;
import java.util.stream.Collectors;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.schedulers.Schedulers;
import js.task.data.local.DbRepository;
import js.task.data.local.db.model.DataModel;
import js.task.data.remote.RetrofitRepository;
import js.task.domain.interfaces.IDataProvider;
import js.task.domain.usecase.model.DomainModel;


public class DataProvider
        implements IDataProvider
{
    private final CompositeDisposable compositeDisposable;
    private final DbRepository localRepository;
    private final RetrofitRepository remoteRepository;

    public DataProvider(
            CompositeDisposable compositeDisposableInject,
            DbRepository localRepositoryInject,
            RetrofitRepository remoteRepositoryInject
    )
    {
        compositeDisposable = compositeDisposableInject;
        localRepository     = localRepositoryInject;
        remoteRepository    = remoteRepositoryInject;
    }

    @Override
    public void download()
    {
        compositeDisposable.addAll(remoteRepository.get()
                .observeOn(Schedulers.io())
                .subscribeOn(Schedulers.io())
                .subscribe(
                localRepository::updateData
        ));
    }

    @Override
    public Observable<List<DomainModel>> getData()
    {
        return localRepository.getData()
                .map(this::mapToDomain);
    }

    private List<DomainModel> mapToDomain(List<DataModel> dataList)
    {
        return dataList.stream()
                .map(data ->
                        new DomainModel(data.getId(), data.getUserName(), data.getImageUrl(), data.getApiName() == null || data.getApiName()
                        .name()
                        .isEmpty() ? "" : data.getApiName()
                        .name()))
                .collect(Collectors.toList());
    }
}

File: data/src/main/java/js/task/data/DataProvider.kt
package js.task.data

import js.task.data.local.DbRepository
import js.task.data.local.db.model.DataModel
import js.task.data.remote.RetrofitRepository
import js.task.domain.interfaces.IDataProvider
import js.task.domain.usecase.model.DomainApiName
import js.task.domain.usecase.model.DomainModel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import javax.inject.Inject


class DataProvider @Inject constructor(
    private val coroutineScope : CoroutineScope = CoroutineScope(Dispatchers.IO),
    private val localRepository : DbRepository,
    private val remoteRepository : RetrofitRepository
) : IDataProvider
{
    override fun download()
    {
        coroutineScope.launch {
            remoteRepository.get()
                .collect {
                    localRepository.updateData(it)
                }
        }
    }

    override fun getData() : Flow<List<DomainModel>>
    {
        return localRepository.getData().map(mapToDomain())
    }

    private fun mapToDomain() : (List<DataModel>) -> List<DomainModel>
    {
        val transform: (List<DataModel>) -> List<DomainModel> = {
            it.map {
                DomainModel(it.id,
                            it.userName,
                            it.imageUrl,
                            DomainApiName.valueOf(it.apiName?.name ?: ""))
            }
        }
        return transform
    }
}